Add conformance profiles NPEP (#138)
kubernetes-sigs/network-policy-api: npep/npep-137-conformance-profiles.md

can run the existing conformance test suite specifying the profiles they support and that will
generate an output that looks like this:

```
    Conformance report:
        apiVersion: policy.networking.k8s.io/v1alpha1
        date: "2023-10-03T08:15:25+02:00"
        implementation:
          contact:
          - "@tssurya"
          organization: ovn-org
          project: OVN-Kubernetes CNI
          url: "https://github.com/ovn-org/ovn-kubernetes"
          version: 0.1.1
        kind: ConformanceReport
        networkPolicyV2APIVersion: v0.1.1
        profiles:
        - core:
            failedTests:
            - AdminNetworkPolicyIngressSCTP
            - AdminNetworkPolicyEgressUDP
            result: failure
            statistics:
              Failed: 2
              Passed: 5
              Skipped: 0
            summary: ""
          name: AdminNetworkPolicy
        - core:
            result: success
            statistics:
              Failed: 0
              Passed: 7
              Skipped: 0
            summary: ""
          name: BaselineAdminNetworkPolicy
```

This can then be uploaded to network-policy-api/conformance/reports/v.x.x/cni-name.yaml by

> opening a PR. That will then be reviwed and approved by mainainers thus recognising the

Reviewer: I'm wondering if we should require implementations to also provide additional information in/alongside with this report, i.e. github action/jenkins or any other CI/CD job definition and run links. As network-policy-api maintainers, we might want to review how the conformance test was run to led to the generation of the report?

Author: oh great point! that way we are having something that maintainers can actually check and verify against. Let me add that.

Author: I am going to call this `VerificationLink`

Author: haha actually `additionalInfo` is best lol, we leave freedom to implementations what form of proof they want to provide

# NPEP-137: Conformance Profiles

> * NOTE: Adapted with love from https://gateway-api.sigs.k8s.io/geps/gep-1709/ (Then

Reviewer 2: NIT add this blurb in the TLDR section not here.

Author: ack moving this..

supports
* Issue: [#137](https://github.com/kubernetes-sigs/network-policy-api/issues/137)
* Status: Implementable

## TLDR

Add conformance profiles for existing conformance tests which implementations
can leverage when running the conformance test suite in their downstream repos
to report the results of the tests back to the Network Policy API project and
receive recognition (eg. "conformance badge")

## Goals

* Add conformance profiles for existing tests that downstream implementations
can subscribe to to run tests associated to the supported API feature sets

* Add a reporting mechanism via a new CRD where conformance results can be
reported back to the Network Policy API project and provide "conformance badges"
via Network Policy official website to recognize these implementations.

* Expand existing conformance testing framework to account for profiles

## Non-Goals

* The first iteration will just have a simple repo under which the report
results will be stored, no automated reporting infrastructure will be built for this now.
Implementations will need to open PRs against network-policy-api repo to upload
the results.

## Introduction

> Currently, the conformance tests are grouped into `StandardCoreFeatures` and

Reviewer 2: nit for another PR, can we just make this `CoreFeatures` and `ExtendedFeatures`?

Author: i actually like this suggestion, I will fix this in this NPEP and also open a new PR to fix the naming in our existing tests..

Author: @astoycos : See https://github.com/kubernetes-sigs/network-policy-api/pull/155

Author: Considering how you have crd/experimental already maybe we call them "ExperimentalFeatures"

`StandardExtendedFeatures`. The features that fall under the `StandardCoreFeatures`

> are a must for the implementations to be conformant while the features that fall
> under `StandardExtendedFeatures` are opt-ins which are not mandatory.

Reviewer 2: ```suggestion
are a required for conformant implementations, while the features that fall
under `StandardExtendedFeatures` are optional and may only be provided by some Implementations.
```

Author: refactored this per suggestion

> In this NPEP, we will add a concept called `named profiles` which will indicate a
> "level of conformance" on a resource level. We are picking profiles on a resource
> level because that is what makes sense for the project as of today. This can be

Reviewer 2: ```suggestion
In this NPEP, we will add a concept called `named profiles` which will indicate a
"level of conformance" on for a given resource. The initial decision to report conformance profiles on a per-resource level  is a choice which may be re-evaluated by the working group in the future.
This can be
```

> changed as the project evolves. Using these `named profiles`, implementations can
> run the conformance tests and prove they satisfy the requirements for each profile
> and receive conformance badges from the Network Policy API project.

Reviewer 2: ```suggestion
changed as the project evolves. Using these `named profiles`, implementations can
run the conformance tests, prove they satisfy the requirements for each profile,
and receive conformance badges from the Network Policy API project.
```

This NPEP aims to provide an API (CRD) for reporting results from these conformance
tests. CLI tooling will also be provided for invoking the conformance tests with
the profile names so that the results can be easily submitted back upstream.

## User-Stories/Use-Cases

Story 1: Receive an API conformance badge

> As a CNI maintainer, I want to receive recognition from Network Policy API repo

Reviewer 2: ```suggestion
As a CNI maintainer, I want to receive recognition from Network Policy API subgroup
```

> that my plugin implements `AdminNetworkPolicy` API. This badge will enable us

Reviewer 2: ```suggestion
that my plugin implements the `AdminNetworkPolicy` API. This badge will enable us
```

to advertise upstream feature support in our plugin easily to our users.

Story 2: Track Implementations for each API resource

As a Network Policy project maintainer, I would like to easily track which
implementations are using the defined API resources in our project so that it
paves way for easier collaboration and feedback loops when changes are introduced
to the API in each release.

## API

We will add a new API resource called `ConformanceReport` which will be at
the center of our test result reporting workflow. The implementors running
the tests can then:

1. Choose a `named profile`

> 2. Add integration for the tests under this profile in their downstream project

Reviewer 2: ```suggestion
2. Integrate the tests required by a given profile in their downstream project
```
Is this what you were trying to say?

Author: yes :D changed it according to your words which make it more clear.

3. Report the results to Network Policy API project using this new API resource
4. Get a conformance badge recognition via our official website

## Profiles

Named Profiles for Network Policy API project will be tied to each API resource.
We will start with two named profiles and expand this as the project evolves.

1. AdminNetworkPolicy
2. BaselineAdminNetworkPolicy

Each of these profiles may have a combination of conformance tests that fall under
`StandardCoreFeatures` and `StandardExtendedFeatures`. Example; if I pick the profile
`AdminNetworkPolicy`, all tests like `AdminNetworkPolicyEgressSCTP` and
`AdminNetworkPolicyPriorityField` fall under the `SupportAdminNetworkPolicy` feature
which is under the `StandardCoreFeatures` subset. So these tests must pass for the
`AdminNetworkPolicy` profile conformance. Whereas tests (we don't have any today)
that fall under `AdminNetworkPolicySameLabels` feature which is under the
`StandardExtendedFeatures` subset are not mandatory for `AdminNetworkPolicy`

> profile conformance.

Reviewer 2: Is there a way to report all of the following:

"I support ONLY the standard features"
"I support All the standard features and some extended"
"I support ALL features"

Author: good question, so the report generates something like this:
```
profiles:
        - core:
            failedTests:
            - AdminNetworkPolicyIngressSCTP
            - AdminNetworkPolicyEgressUDP
            result: failure
            statistics:
              Failed: 2
              Passed: 5
              Skipped: 0
            summary: ""
          name: AdminNetworkPolicy
        - core:
            result: success
            statistics:
              Failed: 0
              Passed: 7
              Skipped: 0
            summary: ""
          name: BaselineAdminNetworkPolicy
       - extended:
           result: success
           statistics:
             Failed: 0
             Passed: 2
            Skipped: 0
           name: AdminNetworkPolicySameLabels
```
from here we can infer does this support ALL features (all pass), ONLY Core (only those will be reported as pass), mix of Core and Extended, and which ones are failing...

Reviewer 2: Sweet! That's what I was looking for :)
